rfcs/45-stageless.md

```diff
@@ -0,0 +1,725 @@
+   ordering_constraints: Graph<SystemId>,
+   // Set configuration must be stored at the schedule level,
+   // rather than attached to specific set instances
+   sets: HashMap<Box<dyn SystemLabel>, SystemSetConfig>,
+   // Used to check if mutation is allowed
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: System,
+   config: SystemConfig,
+}
+```
+
+Raw systems store metadata, cached state, `Local` system parameters and the function pointer to the actual function to be executed:
+
+```rust
+struct System {
+   function: F,
+   // Contains all metadata and state
+   meta: SystemMeta
+   last_change_tick: u32,
+}
+```
+
+System configuration stores ordering dependencies, run criteria and system sets.
+It can be created raw, or stored in association with either system sets or systems.
+
+```rust
+enum OrderingTarget {
+   Label(Box<dyn SystemLabel>),
+   System(SystemId),
+}
+
+struct SystemConfig {
+   strict_ordering_before: Vec<OrderingTarget>,
+   strict_ordering_after: Vec<OrderingTarget>,
+   // The `TypeId` here is the type id of the flushing system
+   flushed_ordering_before: Vec<(TypeId, OrderingTarget)>,
+   flushed_ordering_after: Vec<(TypeId, OrderingTarget)>,
+   run_criteria: Vec<SystemId>,
+   // The combined access requirements of the system and its run criteria
+   joint_access: FilteredAccessSet<ComponentId>,
+   sets: Vec<Box dyn SystemLabel>,
+}
+
+struct SystemSetConfig {
+   system_config: SystemConfig,
+}
+
+```
+
+### System set configuration
+
+When building the dependency graph, all edges (constraints) between labels are expanded into edges between the systems nested within them.
+
+### Run criteria
+
+Run criteria are systems with several key constraints:
+
+- They cannot have system params that give mutable access to data.
+  - Required so we don't have order to them.
+  - Read-only access avoids very surprising side effects.
+- They cannot capture or store local data.
+  - Required to avoid user footguns: local data would be updated even if the guarded system ultimately doesn't run.
+- They must return a `bool`.
+  - Required so we know whether the system should run!
+
+Run criteria are shared.
+They are evaluated a single time during each schedule pass, and the result applies to all systems with that run criteria for the tick.
+
+ When the executor is presented with a ready system, we do the following:
+
+- Check that all the data accessed by the system, its run criteria (RC), and the RC of any sets it is under (that haven't been evaluated yet) is available.
+  - If not, leave the system in the ready queue and move onto the next one.
+- If yes, evaluate the RC of the non-evaluated sets that the system is under in hierarchical order (from the outermost set to the innermost set).
+  - If any set's RC return false, mark all the systems (incl. current one) and sets under that set as completed/evaluated.

```

Comment:
> If any set's RC return false, mark all the systems (incl. current one) and sets under that set as completed/evaluated.

If I understood correctly, this may open the door to a potential bug. Since RCs are systems, I assume the statement “mark all the systems \[...\]” includes them.

This means that a RC (that we'll call “skipped RC”) under a guarded set can be marked as completed/evaluated without actually running.

Recall that a single RC can be used to guard multiple systems in a schedule, but its output is memoized until the end of the schedule run for performance and consistency reasons.

Then, what happens if later in the schedule run, the memoized output of the skipped RC is needed to determine if a system should run or not? The RC didn't run, and we'll probably have a default value for it's output that may be wrong.

Reply:
(This is a fantastic question. Sorry if my answer isn't reflected in the RFC yet.)

In our design, all systems and system sets have unique instances of their RC. None are shared.

RC are evaluated just before the guarded system (or first system in a guarded set) runs. So if a system runs, its conditions and the conditions of all the sets it falls under must have all returned true, and those results are as fresh as possible.

What I mean is they're evaluated just before we would spawn and submit the task that runs the guarded system. All RC are evaluated in the same thread (the one spawning tasks) for perf.

(Our reasoning there is that read-only tests should be super cheap most of the time, so threading would bring relatively significant overhead.)

Reply:
Thanks for the explanation. That definitely eliminates the scenario i described earlier, as all RC are guaranteed to run (i.e. they are never skipped).